### ISSUE TITLE
chooseFloat() and its variants are seriously broken

### ISSUE THREAD

**Issue by jducoeur** [scalatest/scalatest#1473]
I started simplifying the chooseFloat() variants (the way I had already done for the chooseIntegral() ones), and realized that not only was my simplified code broken, but the functions are and always have been severely broken in the presence of edge-case parameters.

A straightforward example of the brokenness: if you `chooseFloat(Float.NegativeInfinity, 100.0.toFloat)`, and the next random number kicked out is, say, 1000.0 (or anything else greater than 100.0), it will return NaN.

If these functions are going to allow for all of the possible edge cases (Min/Max, infinities and NaNs) -- and I should note that I strongly disagree with Bill that the not-really-numbers should be valid in Randomizer (that is what Generator is for) -- then somebody needs to very carefully analyze all the possible function inputs, combined with all the values that might be kicked out by the random-number generation, and make sure that we get sane and reasonably well-distributed results.

This is non-trivial, because Float math isn't designed to cope with infinities and NaNs, and doesn't always play nicely with Min/Max, so the current code tends to break badly if the parameters are edges.  I suspect a bunch of special-case code will be required, and that special-case code may need to vary from function to function because the various numeric types have different legal values.

Note that all of the above is probably true of chooseDouble() and its variants as well, but I haven't tried to play with those.

Note that this *may* be a blocker for 3.1, since it affects what results will be produced for what seeds.  If we require that the results be consistent going forward, then this needs to be fixed now.  (I don't know whether this is a priority or not.)

**Comment by cheeseng** [comment]
@jducoeur @bvenners interesting, imho chooseFloat returning a NaN should be considered a bug, while other edge cases still make sense for being in the range specified by the from and to (which can be Min/Max, infinities) input parameters to the functions, NaN makes no sense here, neither as input parameters or output returned.

I did a quick try on the algorithm to see how the NaN came about, it seems the following will turn into a NaN: 

Float.PositiveInfinity + Float.NegativeInfinity

While I think it is pointless to argue whether the result of it is NaN makes sense (it is implemented that way by JVM anyway), we'll probably need to handle that case in the algo.

I remember seeing flicker related to NaN on and off (quite rare in frequency), but I am glad we probably have found the root cause of it now.

**Comment by cheeseng** [comment]
@bvenners Yes, I was studying your PR, the flicker should be an old one, not latest changes from you or @jducoeur .
